Add a periodic Postges -> Elasticsearch sync task

seanh opened the pull request:
Add a `job` table to the DB for holding annotation sync jobs and a `"job_queue"` service that, when called, consumes a fixed number of jobs from the queue and syncs the annotations from Postgres to Elasticsearch. The number of annotations that the service syncs each time it's called is controlled by the `ES_SYNC_JOB_LIMIT` envvar.

Also add a Celery task that calls the new service. This Celery task is meant to be called by h-periodic on a fixed periodic schedule so that the service will consume jobs from the queue at a fixed rate.

Finally, change the search admin page's "reindex annotations in date range" feature to work by finding the annotation IDs and adding them to the job queue. This means that no matter how many annotations are reindexed, no matter how long the queue grows, they'll only be indexed at a fixed rate by the periodic task.

See also:

* The h-periodic PR that schedules the new Celery task to run periodically: https://github.com/hypothesis/h-periodic/pull/17
* The DB migration to add the `job` table that this PR uses: https://github.com/hypothesis/h/pull/6296

The plan is to use the "reindex annotations in date range" admin page to test this in production and then:

1. Change the other admin pages that do reindexing to also use this job queue
2. Change the annotation create and update APIs to also append jobs to this queue with a delayed `scheduled_at` time
3. Send a separate PR to add queue length monitoring in New Relic

## Testing

1. Reset your DB to get the new table: `make services args='down' && make services db devdata`

2. Create 10,000 annotations that are in your DB but not in Elasticsearch: `tox -qe dev -- sh bin/hypothesis --dev create-annotations --number 10000`

3. Go to http://localhost:5000/search and observe that the annotations aren't visible

4. Log in as `devdata_admin`, go to http://localhost:5000/admin/search, and schedule a reindexing of a time frame

5. Observe that jobs have been added to the job queue: `tox -qqe dockercompose -- exec postgres psql -U postgres -c "select * from job;"`

6. Checkout the [corresponding h-periodic PR](https://github.com/hypothesis/h-periodic/pull/17), run `make dev` in h-periodic, and wait for a minute for the sync-annotations task to be run

7. The first time the task runs you should see output like this in h's terminal:

   ```
   Received task: h.tasks.job_queue.sync_annotations[c0821314-bfe9-465e-ace4-d9286737636f]  
   GET http://localhost:9200/_search [status:200 request:0.040s]
   Syncing 1000 annotations that are missing from Elasticsearch
   POST http://localhost:9200/_bulk [status:200 request:0.132s]
   POST http://localhost:9200/_bulk [status:200 request:0.059s]
   ...
   POST http://localhost:9200/_bulk [status:200 request:0.036s]
   Task h.tasks.job_queue.sync_annotations[c0821314-bfe9-465e-ace4-d9286737636f] succeeded in 3.453727593005169s: None
   ```

8. Go to http://localhost:5000/search and observe that 1000 annotations are now visible

9. Observe the no jobs have been removed from the job queue yet

10. The second time the job runs you should see output like this:

    ```
    Received task: h.tasks.job_queue.sync_annotations[d7fe0e2b-1c9c-4e2c-9e0f-29ad11c3fd07]  
    GET http://localhost:9200/_search [status:200 request:0.077s]
    Deleting 1000 successfully synced annotations from job queue
    Task h.tasks.job_queue.sync_annotations[d7fe0e2b-1c9c-4e2c-9e0f-29ad11c3fd07] succeeded in 0.2429510239744559s: None
    ```

11. Observe that 1000 jobs have now been removed from the job queue

12. If you leave it running for another 18 minutes it will eat through the rest of the job queue at a rate of 1000 jobs every 2 minutes

jon-betts commented on h/storage.py:
```diff
@@ -147,6 +147,8 @@ def create_annotation(request, data, group_service):
     request.db.flush()
 
+    request.db.add(models.EsSyncTask(annotation_id=annotation.id))
+
```
Might want to move this to the `SearchIndexService` as `queue_for_indexing(annotation_id, priority)` or something. That way we can consolidate any smarts about doing it there.

This also means all roads start to lead to the same code, rather than secret connections being made in the background.

seanh replied:
Yeah, that's easily enough done

jon-betts reviewed (commented):
Some random musings

seanh commented on h/models/es_sync_task.py:
```diff
@@ -0,0 +1,19 @@
+from h.db.types import URLSafeUUID
+
+
+class EsSyncTask(Base):
```
The table is a queue, but each individual row in the table is a "task" (or a "message"?). Similar to for example `models.Annotation` rather than `Annotations`

seanh commented on h/models/es_sync_task.py:
```diff
@@ -0,0 +1,19 @@
+    annotation_id = Column(URLSafeUUID, nullable=False)
+
+    #: How many attempts that have been made to complete this task.
+    attempts = Column(Integer, server_default=text("0"))
```
I think I'm gonna leave out priority for now. It isn't needed until we have messages being added to the queue for multiple different reasons, and when we do come to do that I'm not 100% sure that a simple priority int is going to be enough. If we just have integer priorities and the task consumes messages high-priority first then the lower-priority messages might never get consumed since the high-priority ones are coming in all the time. So we might want something like multiple periodic tasks, one for each priority. We might end up still wanting an integer priority column in the end but I think we should add it at the time as we'll know better then

seanh commented on h/tasks/es_sync.py:
```diff
@@ -0,0 +1,107 @@
+                    log.info(
+                        f"Deleting a message from the queue because it failed to sync to the search index too many times: {message}"
+                    )
+                    db.delete(message)
```
This actually both indexes the message and deletes it from the queue in one run of the task, which is wrong. It makes the logging confusing. The outer loop should be iterating over `messages` not over `annotations_from_db`. Need to fix that

seanh commented on h/tasks/es_sync.py:
```diff
@@ -0,0 +1,107 @@
+
+
+def get_annotations_from_db(db, messages):
+    # TODO: Filter out very recent messages (need to add date to model).
```
This TODO is in the wrong place

seanh commented on h/tasks/es_sync.py:
```diff
@@ -0,0 +1,107 @@
+            .limit(os.environ.get("ES_SYNC_TASK_LIMIT", 1000))
+            .with_for_update(nowait=True)
+            .all()
+        )
```
This query also needs to filter out very recent messages (say added to the queue in the last 30 secs?) so that the periodic task doesn't fight with web requests. That'll require adding a `time_added` column to the queue

seanh commented on h/config.py:
```diff
@@ -100,6 +100,7 @@ def configure(environ=None, settings=None):
     # *not* toggle functionality based on this value. It is intended as a
     # label only.
     settings_manager.set("h.env", "ENV")
+    settings_manager.set("h.es_sync_job_limit", "ES_SYNC_JOB_LIMIT", default=1000)
```
So we'll have two ways to tune the speed of reindexing in production:

1. We can change the `ES_SYNC_JOB_LIMIT`, which is how many annotations the service will check per run
2. We can change the schedule in h-periodic, which is how frequently the service gets run

seanh commented on h/models/job.py:
```diff
@@ -0,0 +1,32 @@
+        JSONB,
+        server_default=text("'{}'::jsonb"),
+        nullable=False,
+    )
```
`tag` is gonna be used to distinguish jobs added by admin pages from those added by the API. For example if someone uses an admin page to add a lot of jobs to the queue and we want to remove them, this will make it easy to delete only the jobs added by the admin page.

`scheduled_at` is gonna be used by the API to schedule jobs not to be processed until a minute after the API call. (Whereas jobs scheduled by admin pages can be processed immediately.)

`enqueued_at` is for sorting.

seanh replied:
I think we might want to add a `type` column (string) even though, currently, all rows would have the value `"sync_annotation"`. I think a `type` column is the one thing that's missing that would be needed to potentially reuse the job queue for a second type of job in the future, without having to do a data migration to add a `type` to existing rows.

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,170 @@
+        This method is meant to be run periodically. It's called by a Celery
+        task that's scheduled periodically by h-periodic.
+
+        Each time it runs the method consumes a fixed number of sync annotation
+        jobs from the queue and copies those annotations from Postgres to
+        Elasticsearch (if they aren't in Elasticsearch already).
```
TODO: Improve this docstring

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,170 @@
+
+            if not annotation_from_db:
+                deleted.append(job)
+            elif annotation_from_db.deleted:
+                deleted.append(job)
```
This could be simplified by changing the SQL query below to filter out annotations marked as deleted in the first place

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,170 @@
+            elif not annotation_from_es:
+                to_index.append(job)
+            elif annotation_from_es["updated"] != annotation_from_db.updated:
+                to_index.append(job)
```
Might want to keep separate `missing_from_es` and `outdated_in_es` lists so that we can log different messages for the different reasons

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,170 @@
+            )
+            .order_by(Job.enqueued_at)
+            .limit(self._limit)
+            .with_for_update(skip_locked=True)
```
For `FOR UPDATE` and `SKIP_LOCKED` see:

* https://www.pgcon.org/2016/schedule/attachments/414_queues-pgcon-2016.pdf
* https://www.depesz.com/2014/10/10/waiting-for-9-5-implement-skip-locked-for-row-level-locks/
* https://www.pgcasts.com/episodes/the-skip-locked-feature-in-postgres-9-5
* https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/

Basically this uses Postgres's `SELECT FOR UPDATE` to lock all the rows that we're selecting for the duration of our transaction. Then it uses Postgres's `SKIP LOCKED` to skip over any rows locked by another transaction. This means that if two instances of the service get run at the same time, they won't attempt to process the same jobs.

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,193 @@
+            )
+
+        if missing_from_es or outdated_in_es:
+            self._batch_indexer.index(
+                [
+                    job.kwargs["annotation_id"]
+                    for job in missing_from_es + outdated_in_es
+                ]
+            )
```
Should this call to `BatchIndexer.index()` have any error handling? Or should we just let the Celery worker crash if there's an error?

seanh replied:
Should this fall back to plain Elasticsearch indexing (non-batch) if the number of annotations to be indexed is small? Currently this will use the batch indexer even to index one annotation, but maybe that's fine

jon-betts reviewed (changes requested):
Given the size of this, I think this will be a review in stages, so I'll fess up right now to not having gone through the tests yet. There's also going to be a lot of feedback here, not because this is bad, but again, just because it's big.

In no particular order:

**About the algorithm**

Seems pretty sound to me. The only things I'd say are:

 * A priority column would allow us to handle repeat offenders quite gracefully, currently I think they will pile up at the start of the queue
 * You can do something like that by pushing back the "schedule at" time, but it would be hard to get the same nice behavior, rather than just causing a problem in X minutes.
 * This is the only use of error handling around the `BulkIndexer` I can see, so if we're not going to do that, then maybe this is fine

**About the layout**

 * The name `JobQueue` gives me pause ...
 * ... _however_ I think if we start viewing this as essentially private to the `search_index` service things make more sense
 * It would become `h.services.search_index.job_queue` which makes total sense
  * The previous work was starting to gather indexing logic towards the `search_index` service. In it's current state I feel this is going the other way, and starting to spread it out again. If we popped it in with the `search_index` then I think it would continue the good trend
 * The tasks would then gather naturally too
 * This would cleanly lead to a future where you go through `search_index` to get things done, and all the bulk handling, or defered checking etc. would be applied centrally.

**About the code**

 * I think there's good scope to simplify the main loop by changing the lists to be outcome oriented rather than situation oriented (i.e. `update_required` rather than `missing_in_es` etc.)
 * We're going to be issuing a lot of deletes to jobs hopefully, so it's probably worth the time to make sure that happens in one query
 * I think the logic which has gone to the admin view would be better in the search index / job queue service. This would also point the way towards where I see this going where you could say something like 'search_index.index_these(..., delay=True)` and it would queue them for you, or do them now as appropriate.
 * We could use the same mechanism we did in LMS for adding `__repr__` methods to all the DB models

**About the style**

 * The docstrings are too long for my taste
 * Some of this is because they are a bit wordy, but most of it is because they are documenting the situation/system rather than the class they are in
 * This means if you just want to know what a function does there's a lot of extraneous content, and it pushed the code a long way down
 * I think this long format discussion around the usage of the class might be better at the top of the file or something

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,193 @@
+        missing_from_db = []
+        missing_from_es = []
+        outdated_in_es = []
+        up_to_date_in_es = []
```
It's necessary to keep these four separate lists because each list has a different log message associated with it, which I think we want for debugging and monitoring what the process is up to

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,193 @@
+        if not jobs:
+            return
+
+        annotation_ids = [job.kwargs["annotation_id"] for job in jobs]
```
Hmm, possibly. I guess they are in order and would lose their order if put into a set. But since they end up getting treated in bulk anyway that shouldn't matter. And I think the same annotation ID _could_ appear twice in a block of N jobs, which could result in us trying to bulk index that annotation into Elasticsearch twice in one call or something. As well as passing the same ID twice to the Postgres and Elastic queries. So a set might be a good idea just to remove that possibly confusing or broken edge case entirely

seanh replied:
Changed it to use sets in a few places: https://github.com/hypothesis/h/pull/6265/commits/7790b35f721e921cc18f7aa971af2d3f310f373e

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,193 @@
+        outdated_in_es = []
+        up_to_date_in_es = []
+
+        for job in jobs:
```
I think it might be an over-optimisation. It may be easier to ignore that optimisation for the sake of keeping the code simpler. On the other hand it's probably do-able

jon-betts replied:
:+1:

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,193 @@
+    return JobQueue(
+        db=request.db,
+        es=request.es,
+        batch_indexer=BatchIndexer(request.db, request.es, request),
```
Yeah it feels wrong that the job queue's service factory has to now how to instantiate the `BatchIndexer`. It feels like the `BatchIndexer` should be wrapped in its own service with its own factory that knows how to instantiate it, then the job queue's factory can just get it with `request.find_service(..)`. And there'd only be one place in the code that knows how to construct a `BatchIndexer`. So yeah folding the batch indexer into the `search_index` service would achieve that

jon-betts replied:
Although if the search service becomes a front for this one, then it makes sense that all these things all know about each other.

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,193 @@
+                missing_from_db.append(job)
+            elif not annotation_from_es:
+                missing_from_es.append(job)
+            elif annotation_from_es["updated"] != annotation_from_db.updated:
```
Yeah I guess that would make the word "outdated" incorrect. Could rename `outdated_in_es` to `different_in_es`?

I think the actual logic is fine--we're doing a one-way sync from Postgres to Elastic so even if the date is newer in Elastic (somehow) it's still correct to overwrite what's in Elastic with what's from Postgres. So maybe just the variable naming is wrong.

seanh replied:
Renamed this: https://github.com/hypothesis/h/pull/6265/commits/df6ca0d84111c0d6d566e3ea1903c23ab027d916

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,193 @@
+        return {
+            annotation.id: annotation
+            for annotation in self._db.query(
+                Annotation.id, Annotation.updated, Annotation.deleted
```
This has been removed

seanh commented on h/tasks/job_queue.py:
```diff
@@ -0,0 +1,6 @@
+
+@celery.task
+def sync_annotations():
+    celery.request.find_service(name="job_queue").sync_annotations()
```
It depends what you see the job queue as.

I see it as a general-purpose job queue that has different properties from the Celery task queue (it's transactional with the DB etc). It just so happens to only be used for sync-annotation jobs currently. And may only ever be used for those jobs. But it could be used in the future for some other type of job that has nothing to do with annotations or the search index. So I see the job queue as its own separate thing, not part of the `search_index` service.

The Celery task for consuming from the job queue though (`JobQueue.sync_annotations()`) could possibly call a `search_index` service method to do the actual syncing of a batch of annotations from Postgres to Elastic? Since syncing annotations is a search index thing. The split of responsibilities could be awkward though as reading a batch of jobs from the job queue and possibly deleting those jobs from the queue seems like a job queue thing, whereas reading annotation IDs from Postgres and Elastic and possibly syncing them into Elastic is maybe a search index thing. So it seems that the algorithm that's currently in one method would have to be split up into two places

jon-betts replied:
I think it's pretty clear right now it's job is 1:1 with indexing. In the event that we ever did something else with it, I think that would be the time to think about disentangling it.

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,193 @@
+            )
+
+        for job in missing_from_db + up_to_date_in_es:
+            self._db.delete(job)
```
I thought this _was_ batched into one query by SQLAlchemy, but I can check that. I think the call to `delete()` just marks the ORM object as deleted in memory, and then when the SQLAlchemy session is flushed or committed SQLAlchemy batches all the in-memory ORM changes into as few SQL queries as possible. I'll see if I can double check what SQL queries it's actually emitting though.

The SO answer is about something else, I think. He's asking about how to reduce a single `SELECT` followed by a single `DELETE` (of the same rows that were selected) into just a `DELETE`. (Rather than how to reduce multiple `DELETE`s into one `DELETE`, which I _think_ SQLAlchemy just does.)

jon-betts replied:
Fair enough!

seanh replied:
I've tested this and SQLAlchemy does batch them for us

seanh commented:
> A priority column would allow us to handle repeat offenders quite gracefully, currently I think they will pile up at the start of the queue

I'd like to keep it simple and avoid the `priority` column and decrementing priority for now. Instead, just go with queue size monitoring and alerting in New Relic (to be added in a future commit). So that at least if a _lot_ of repeat offenders get added to the queue an alarm will go off.

That doesn't prevent a few repeat offenders from getting added to the queue and sitting there forever though. Maybe a way to deal with that could be to expire _really_ old jobs from the queue? Like, just delete a job that is older than one month? Either based on the `enqueued_at` date or, probably better as it's more flexible, based on a separate `expires_at` column that defaults to one month after the row is enqueued?

jon-betts commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,193 @@
+            else:
+                up_to_date_in_es.append(job)
+
+        if missing_from_db:
+            logger.info(
+                f"Deleting {len(missing_from_db)} sync annotation jobs because their annotations have been deleted from the DB"
+            )
+        if up_to_date_in_es:
+            logger.info(
+                f"Deleting {len(up_to_date_in_es)} successfully synced annotations from job queue"
+            )
+
+        for job in missing_from_db + up_to_date_in_es:
+            self._db.delete(job)
```
That feels like something you probably care about right now, but probably don't overall. You could keep a separate tally in a `collections.Counter` and just print it out instead

seanh commented on h/models/job.py:
```diff
@@ -0,0 +1,33 @@
+                "kwargs",
+            )
+        )
+        return f"Job({attrs})"
```
I've just removed the `__repr__` for now: https://github.com/hypothesis/h/pull/6265/commits/aa4be364e0ce222794cff83cffc7b36552ca30db It isn't really needed (I was just using it for debugging at one point) and I want to keep this as slim as possible

seanh commented on h/services/job_queue.py:
```diff
@@ -0,0 +1,193 @@
+        If the DB somehow contains an annotation that always fails to index
+        into Elasticsearch, then a job for that annotation will never be
+        removed from the queue! As far as I know it isn't possible for the DB
+        to contain such an annotation.
```
I've edited these docstrings: https://github.com/hypothesis/h/pull/6265/commits/279dce1f1af5553aa1c416bbe7beb605e4ff760a

jon-betts commented on h/services/search_index/service.py:
```diff
@@ -107,6 +131,9 @@ def handle_annotation_event(self, event, synchronous=False):
         # Either the synchronous method was disabled, or failed...
         return async_task.delay(event.annotation_id)
 
+    def sync(self):
```
Doc string?

jon-betts commented on h/templates/admin/search.html.jinja2:
```diff
@@ -7,9 +7,6 @@
   </p>
-  <p>
-    The background worker will refuse to process any items after 250,000.
-  </p>
```
As we no longer pass in 'indexing' or 'task_id' a tonne of this template can evaporate. We don't need the indexing warnings or anything as it doesn't apply any more

jon-betts commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+                    tag="test_tag",
+                    kwargs={"annotation_id": "test_annotation_id"},
+                )
+            ),
```
Good to see `h-matchers` getting a work out!

jon-betts commented on h/services/search_index/_queue.py:
```diff
@@ -0,0 +1,146 @@
+        """Queue an annotation to be synced to Elasticsearch."""
+        self.add_all([annotation_id], tag, scheduled_at)
+
+    def add_all(self, annotation_ids, tag, scheduled_at=None):
```
Would `scheduled_at` be better as a delay (from now) instead?

seanh replied:
Hmm.

`scheduled_at` needs to be a datetime not a duration in the DB because we filter by it--if it was a duration on top of `enqueued_at` in the DB then Postgres would have to calculate `enqueued_at` + `scheduled_at` for every row in the DB every time we want to retrieve jobs.

So we'd have to implement this by having the Python service accept a duration and transform that into a datetime for the DB. Something like:

```python
def add_all(..., scheduled_at=None):
    scheduled_at = utcnow() + scheduled_at
    ...
```

I'm calculating the date from `utcnow()` not from `enqueued_at`. The Python code can't do `scheduled_at = enqueued_at + scheduled_at` because `enqueued_at` is not computed by Postgres until it commits the rows (`server_default`), it's not available to Python until after the rows have been committed. Except maybe if you flush the DB first but that's gonna be awkward and inefficient especially when adding a lot of rows. Calculating from `utcnow()` should be fine though I'd think.

seanh replied:
:+1: Done

jon-betts replied:
Gah, I really should have said, if we're going to change the type the argument name should change too.

In the DB it's `scheduled_at` but here it's probably `scheduled_in` until it's added to now.

seanh replied:
Done: https://github.com/hypothesis/h/pull/6265/commits/3f9996cd5f153e98e2cb6f609533c3ad626750af I just couldn't think of another name for it, but `schedule_in` works

seanh commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+                    for annotation in annotations[:limit]
+                ],
+            },
+        }
```
This should probably be changed to use the real Elasticsearch instead of mocking its return values

jon-betts commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+        queue.add_all(
+            annotation_ids,
+            tag="test",
+            scheduled_at=one_week_from_now,
```
If you knock the trailing commas off of all these statements and the signatures, and format again significant vertical space can be reclaimed in this file (about 75 lines)

seanh replied:
:+1: Done

jon-betts commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+        annotation_ids,
+        caplog,
+        db_session,
+        queue,
```
Here too etc.

seanh replied:
:+1: Done

jon-betts commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+                "Syncing 1 annotations that are missing from Elasticsearch",
+            )
+        ]
+        batch_indexer.index.assert_called_once_with(
+            Any.list.containing([annotation_ids[0]]).only()
+        )
```
```suggestion
        batch_indexer.index.assert_called_once_with([annotation_ids[0]])
```

When there's only one item exactly a plain old list match is probably the ticket.

seanh replied:
:+1: Done

jon-betts commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+        batch_indexer.index.assert_not_called()
+
+
+one_week_from_now = datetime.datetime.utcnow() + datetime.timedelta(weeks=1)
+
+
+five_minutes_ago = datetime.datetime.utcnow() - datetime.timedelta(minutes=5)
+
+
+limit = 10
```
```suggestion
LIMIT = 10
ONE_WEEK_FROM_NOW = datetime.datetime.utcnow() + datetime.timedelta(weeks=1)
FIVE_MINUTES_AGO = datetime.datetime.utcnow() - datetime.timedelta(minutes=5)
```

These are super confusing. With fixtures and such there are a lot of interestingly scoped things going about. Upper casing them suggests to the reader they are just constants. As such I'd also stick them at the top of the file.

seanh replied:
Yeah, these should be capitalised, done.

Let's keep them at the bottom of the file though -- our style is that constants, helper functions, fixtures etc go at the bottom of the file or class whenever possible. The thinking is that the actual tests (or source code for source files) is the interesting thing that I want to get to first, I don't want to scroll past a bunch of constants or all the private helper functions etc before getting to the actual code. Those can be hidden away at the bottom where you can go and look at them if you need to know what they are, but otherwise they don't get in your way. It also keeps the constants together with the fixtures and helpers which're already at the bottom.

jon-betts commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+
+@pytest.fixture
+def annotations(factories):
+    return factories.Annotation.create_batch(size=20)
```
Instead of applying `LIMIT` to every use of annotations, we could make more than we need here (for the noise) and then only return `annotations[:LIMIT]`.

If we did that, it's possible that `LIMIT` is actually private to this function

seanh replied:
I don't think this would work, for example `test_it_ignores_jobs_beyond_limit()` needs to enqueue jobs for all 20 annotations (not just the ones up to `LIMIT`) and then needs to assert that the annotations _beyond_ `LIMIT` weren't indexed.

jon-betts replied:
Ok, well something along the lines of `many_annotations` which has all of them and `annotations` which returns the first LIMIT of them? This complicates all tests for one or two as it stands.

seanh replied:
:+1: Done https://github.com/hypothesis/h/pull/6265/commits/e75d849b98ad2523624b05647a58cc7c633db628

jon-betts commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+
+@pytest.fixture
+def annotation_ids(annotations):
+    return [annotation.id for annotation in annotations]
```
The two annotation fixtures here are actually private to the second test, which means `LIMIT` is too, and could be a class variable.

seanh replied:
Moved these into the class. `LIMIT` needs to remain outside though because it's used by the `queue` fixture

jon-betts commented on h/services/search_index/service_factory.py:
```diff
@@ -9,4 +11,10 @@ def factory(_context, request):
+            es=request.es,
+            batch_indexer=BatchIndexer(request.db, request.es, request),
+            limit=request.registry.settings["h.es_sync_job_limit"],
+        ),
```
I'm not sure we make much use of the composition here. It could be private I think, but also... does it matter? Makes this slightly harder to mock as we have to prepare a pretend queue too, rather than letting mock make one for us.

seanh replied:
Keeps the ugly stuff (constructing the `BatchIndexer`) out of the service itself, though

jon-betts reviewed (changes requested):
It's nice to see this pulling together into the same place, but I think there is a way to go. The docstrings are much more focused and have less of a OMG factor.

There are two major things I can spot here:

 * We removed some variable output in the admin search page, but none of the template content that is powered by it. There's significant cruft in there now that doesn't do anything
 * The queue tests are very long, and I think a lot of it can be fixed. I had a go locally and managed to knock 150+ lines out of it (~40%)

Some smaller things:

 * There's a test specifically for tasks which just wrap Search Index, I've added a commit for this
 * I think the scheduled at argument to `add()` (not the DB column) would probably be nicer as an offset, otherwise all consumers have to calculate it for themselves. Most things happen relative to `now`.

**About the queue tests**

The queue tests are much longer than they need to be for a few reasons:
   * They would benefit from a method to setup the `es` return value (3 usages)
   * Many unnecessary return lines possibly as a result of black + trailing commas
   * A few matcher tricks here and there can cut down on vertical space
   * We are testing log output and I strongly believe we shouldn't be
   * I had a go at this and cut about 157 lines (43%) from the file, making it much easier to read

As a separate point, we are testing log output and I strongly believe we shouldn't be. I really can't over state this enough. It makes the tests + code more brittle, and actually doesn't prove the situation it's trying to. Testing log messages is 99.9% of the time an anti-pattern. This is in the majority.

I suspect the use of the logs in the tests was also the source of some of the trepidation about simplifying the core algorithm. If these were removed the main code could also be simplified as suggested before.

A few other suggestions in there like maybe eliminating the `limit` variable, or certainly putting it in upper case (this really confused me about where it was coming from). There is at least one place where we use an order invariant matcher for a list of size one, which isn't needed. Some other stuff I forget.

seanh commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+                    for annotation in annotations[:limit]
+                ],
+            },
+        }
```
I'm hoping the code will be simpler when I change it to use the real Elasticsearch instead of mocking it

seanh replied:
I've changed this to use the real Elasticsearch, which gets rid of these completely

seanh commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+                    ),
+                },
+            ),
+        )
```
I've changed this to use the real Elasticsearch, which gets rid of these completely

seanh commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,364 @@
+                Any(),
+                "Deleting 10 sync annotation jobs because their annotations have been deleted from the DB",
+            )
+        ]
```
I think for monitoring and debugging purposes in production and dev I do want it to log _why_ it thinks it's deleting or indexing annotations, not just _that_ it is. And I think I do want tests for that logging -- I think the logging and its correctness is a significant part of the behaviour that I want from the method. Logging is going to be the main way that we test and monitor this job in production.

The tests do cover that the logging isn't lying. For example they set up a situation in which some jobs' annotations have been deleted from the DB. Then they test that the method both deleted the jobs from the queue _and_ logged that it was deleting them because their annotations were deleted from the DB, and didn't log anything else. So the test covers that the right log messages are logged at the right times. This also helps to make sure that the test set up is correct, and that the test hasn't accidentally set up some other situation that caused the method to delete the jobs for some other reason.

If you try breaking the logging you'll see that any breakage will cause a test to fail. E.g. change it to fail to log one of the log messages, or log the wrong message in the wrong circumstance, etc. The tests have it all covered.

I've changed the `caplog` fixture to only capture log messages from the module under test, so that if anything else that gets called (e.g. elasticsearch etc) logs anything that won't affect these tests.

I'm not concerned about changes to log messages causing tests to fail: it's unlikely that we'll want to change log messages in the future, and if we do the trivial test change will only take two seconds to make. Log message ordering shouldn't break any tests because each test only triggers one log message. It seems like the worst that can happen here is a trivial test breakage could cost a couple of seconds, and that seems unlikely to happen.

jon-betts replied:
There are a few different things mixed together here, both conceptually and in the tests.

### It's not clear in the code if we are testing the logs or using the logs to test

The first thing is it isn't clear if the assertions about the logs are intended to test the logging behavior, or intended to prove the behavior under test. Without you explaining this above it's not clear which intention you have. This is why I said the the logs can lie. Assuming that the code is wrong (which is what tests are about) we can't rely on the logging output to determine what it's doing, it's very circular.

It doesn't sound like that's what you were doing here, so that point is a bit moot, but the confusion is real.

### Fixing logging behaviour in tests increases costs, and makes adding/removing logging less desirable 

Here is a quick test. Add one line of logging to the top of the method and then run the tests. For me 9 of them fail. 

It's not at all clear why, and it causes the undesirable behaviour we were discussing with functional tests where you do one thing and a massive number of things blow up. (Suggestions on how to fix this below!) It also means that when you add a new line, presumably you're expected to add logging for it. It raises the costs all round.

This is because the tests don't just assert the content of the message  they also assert the position of the log message in the stream of logs. You can fix some of this by doing this instead:

``` 
        assert (
            "Syncing 1 annotations that are missing from Elasticsearch"
            in caplog.record_tuples[0]
        )
```

But this still leaves some lines asserting we don't log in certain situations. The point here is, we are locking down functionality that isn't is not, in the future worlds of what this code could be, (imho) important at all.

Is it mandatory that this code _never_ logs in certain situations, or is that just how it happens to behave right now? It is absolutely mandatory that it clears completed jobs etc., but I feel **these tests are enforcing non-functional requirements**.

### I'm not sold that this is how we are going to monitor this

I think this is how you're going to monitor it when you deploy it and look at it today and tomorrow. I think this is also useful debugging info going forward, but this isn't a monitoring solution.

If you want that we should be logging metrics somewhere. Looking for random things in Papertrail isn't how we do this. Which leads to...

## Assuming we're doing this, lets think about how to do it best

This log testing is currently testing two separate behaviours:

 * Collecting information about what situations we discovered to give insight about what we did and why
 * Testing a particular serialisation of that information

When you see assertions made about logging there are generally two questions to ask:

 * Are we testing behavior that doesn't matter?
 * If the behavior matters, how can I make this part of the system rather than a side effect?

Lets assume this is important and think about how we could structure this differently. I'd suggest:

 * Collect and return the stats in question
 * Test this in the individual methods
 * Have a separate test that checks this gets serialised as we'd expect

I've added an example of what that might look like here: https://github.com/hypothesis/h/pull/6309

With a setup like that we can check the counts returned in all of the separate unit tests, and then have one test like `def test_it_logs_the_counts()` which is the only one that would care about the exact serialisation.

This would also be a nice jumping off point if we do want to send these stats to NewRelic or some other collection service.

seanh replied:
* Using a counter does simplify the code quite a bit (only one call to `log.info()`) so I've done that: https://github.com/hypothesis/h/pull/6265/commits/9adf180b71806a16f642d679eef644354d10b5b8

* That commit also makes it so that changing the wording of the log messages will no longer break the tests

* I've also tweaked the assertions so that they don't break if more log lines are added to the method: https://github.com/hypothesis/h/pull/6265/commits/6df62d18704f278fceb9b7f6853aeda2adabc004

* I think it's clear enough that we're testing the logs rather than using the logs to test. Every time we make an assertion about something logged it's right next to an assertion about what actually happened, e.g.:

  ```python
  def test_if_the_annotation_is_missing_from_Elastic_it_indexes_it(self, ...):
      ...

      assert str({MISSING: 10}) in caplog.text
      batch_indexer.index.assert_called_once_with(Any.list.containing(annotation_ids).only())
  ```

* We seem to disagree on whether the logging in this particular function is an actual functional requirement or an incidental side-effect. I consider it to be a real functional requirement and I absolutely am intending to use this to monitor and debug this job in production. These log messages are going to be the only way that we can see what this function is up to -- how many jobs it's deleting and how many annotations it's syncing and why. That's going to be useful for initial testing of the job in production and over its first few weeks of deployment, and also for monitoring the job when it's recovering for us after an outage, or when a big reindex has been scheduled, etc. We could send metrics to New Relic and have a dashboard of graphs instead but I'm not planning to do that work--logging is the cheap-and-easy version. The only thing I'm planning to add to New Relic is basic queue size monitoring and a queue size alarm.

* I'd prefer to just use pytest's builtin `caplog` fixture for this. It's provided by pytest for this purpose, and it's pretty simple (`assert str({UP_TO_DATE: 10}) in caplog.text` etc). I think we should just use this and get on with it. It isn't really necessary here to separate information collected (the counts themselves) from serialization (the logged strings) and doing so would mean I'd have to write more tests and that the tests would be less direct, as there would have to be tests asserting about the returned counts and then separate tests asserting that the returned counts are also logged.

  Instead of using `caplog` we could patch `logger` and then the tests could do things like `logger.info.assert_called_with(expected_counts_dict)`. Rather than having the method return the counts dict for us to assert against. That way there's doesn't have to be a separate "and it also passes the counts that it returns to `logger.info()`" test. But honestly I think pylint has done this for us with `caplog`, and `assert str({UP_TO_DATE: 10}) in caplog.text` isn't much different from `logger.info.assert_called_with({UP_TO_DATE: 10})` (and it's actually more robust because it doesn't care what log level was used).

seanh commented:
Note to self - I want to make a few more changes to this:

- [ ] Remove the `ES_SYNC_JOB_LIMIT` envvar. Instead have the `limit` be an argument sent by h-periodic to the Celery task. No need for an envvar / setting, and h-periodic can then call the task at different intervals with different limits if it once. This also keeps everything about how the task is called (at what interval(s) and with what limit(s)) together in one place in h-periodic's beat schedule

- [ ] A `queue` column. `queue` will be a parameter that h-periodic sends to the Celery task, and the service will filter in only the jobs that match the given queue. The API will then use the `"api"` queue and the admin pages will use an `"admin"` queue and h-periodic will schedule the Celery task twice once for the `"api"` queue and once for `"admin"` (with different periods and `limit`'s if we want).

  This is instead of using the `tag` column for filtering. Let `tag` be what it is--just a piece of metadata not actually used for anything. This also avoids the need to filter multiple tags in and out, we can do simple exact single string matching on `queue` because that's all `queue` is used for.

- [ ] Add a `priority` (int) column and enable the user to optionally enter a custom priority when scheduling reindexing on admin pages. This is so that, if I schedule a massive reindex of a huge time frame and then realise that I want to quickly reindex a small timeframe without waiting for that to finish, I can add the second reindex with a higher priority. Jobs will always be sorted by priority before sorting by `enqueued_at`, and `priority` will default to `0`. The priority of a job never needs to change once the job is enqueued -- the combination of simple `queue` (filtering) and `priority` (sorting) columns should give us all the flexibility we need

jon-betts commented on h/templates/admin/search.html.jinja2:
```diff
@@ -27,36 +18,25 @@
-        </div>
-      </form>
-    {% endcall %}
-  {% endif %}
```
:+1:

jon-betts commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,255 @@
+ONE_WEEK = datetime_.timedelta(weeks=1)
+
+
+MINUS_FIVE_MINUTES = datetime_.timedelta(minutes=-5)
```
I think fixtures go at the bottom, but constants generally go to the top. The double spacing is a touch odd too.

jon-betts commented on tests/h/services/search_index/_queue_test.py:
```diff
@@ -0,0 +1,255 @@
+                    body=body,
+                    id=annotation.id,
+                )
+            es_client.conn.indices.refresh(index=es_client.index)
```
This is a lot nicer. I think we can take a cue from the search tests about how to use the `SearchIndex` service to do this for us instead of writing it again. 

It's nice because it's shorter, but it's also nice because it means we run the real indexing code: https://github.com/hypothesis/h/blob/master/tests/h/search/conftest.py#L46

seanh replied:
This can't easily use the real `SearchIndexService` because it supports overriding the `updated` date when indexing annotations:

https://github.com/hypothesis/h/blob/e75d849b98ad2523624b05647a58cc7c633db628/tests/h/services/search_index/_queue_test.py#L209-L221

One of the tests uses this to create outdated annotations in Elastic:

https://github.com/hypothesis/h/blob/e75d849b98ad2523624b05647a58cc7c633db628/tests/h/services/search_index/_queue_test.py#L130-L133

We'd have to find another way of adding outdated annotations into Elastic if we wanted it to go through `SearchIndexService.add_annotation()` instead

seanh replied:
I found a way to do this: https://github.com/hypothesis/h/pull/6265/commits/8acf41ce374add85923fa87c4c7f43cf49b6d32a

This does unfortunately mean that the `Queue` unit tests now depend on the real `SearchIndexService`, `AnnotationModerationService`, `NIPSAService` and `AnnotationSearchIndexPresenter` but :man_shrugging: That's everything you need to index something. It's probably more of a code design issue than a tests issue: to add an annotation to the DB you only need the `models.Annotation` class (and possibly some dependent model classes). There should probably be a similar model class for an annotation in the search index. IIRC Elasticsearch's official high-level Python client works in this way

jon-betts reviewed (changes requested):
I think we're nearly there but for a few things. Wee ones first:

 * If the argument to `add()` is delay it should probably be called `schedule_in` which becomes `scheduled_at` only after `now()` is added
 * Using real ES is better in the tests for sure, but there's already a fixture for indexing annotations in `h.search.conftest` which I think we can look to, which is shorter and uses the real service

I have some ideas on the way forward about log messages and testing. The obvious one is don't. This is still my best advice.

If we do want to keep it though, I think we can get what we want here and minimise the brittleness and indirectness. My suggestion for this is:

 * Count the things you want to report on separately from how the algorithm works (this allows simpler code in places too) and return these counts
 * Make assertions about these returned counts only in the majority of tests
 * Have a single test that checks we serialise those counts as expected
 * Assert the minimum you can in that test about the format while still testing (i.e. `str(counts) in message` or something)

I made an example of what the code might look like here: https://github.com/hypothesis/h/pull/6309/files#diff-9c8ecb22cb7cb079f05607290725147c

seanh commented:
@jon-betts This is ready for review again. Summary of recent changes:

* Squashed the history

* Replaced the UUID primary key on the `Job` table with an auto-incrementing integer sequence:

  https://github.com/hypothesis/h/blob/f5db917a22ddf98bd154e0bf59c5bdb316f5bba0/h/models/job.py#L41

  I think simple ints are faster than UUIDs (though I doubt it makes much difference) and this removes the need to choose what version of UUID to use. The reason for using UUIDs was that they never run out, whereas we could eventually run out of ints. I've addressed this by passing `cycle=True` so that the sequence goes back to the start once it runs out of ints.

* Added a `priority` column:

  https://github.com/hypothesis/h/blob/f5db917a22ddf98bd154e0bf59c5bdb316f5bba0/h/models/job.py#L48

  The `priority` is always chosen automatically: jobs that're added one at a time get highest priority whereas jobs that're added in large batches get lower priorities:

  https://github.com/hypothesis/h/blob/6df62d18704f278fceb9b7f6853aeda2adabc004/h/services/search_index/_queue.py#L30-L34

  Jobs are sorted by priority then by `enqueued_at` when fetching jobs from the queue:

  https://github.com/hypothesis/h/blob/6df62d18704f278fceb9b7f6853aeda2adabc004/h/services/search_index/_queue.py#L112-L120

  This means that, for example, job's that're reindexing the entire DB will get lower priority than jobs that're NIPSA'ing or renaming a user, which will in turn get lower priority than single-annotation jobs created by the API.

  I think this makes the admin pages safe to use. The worst you can do is make the queue really, really long, but even if you do that:

  * You won't increase system load or create any kind of degradation. The single periodic sync task will just continue eating through the queue at its top speed, no faster.
  * You won't hold up sync annotation jobs added automatically by the API, since those are added with the highest priority (`1`) whereas any big batch jobs will get a much lower priority
  * You won't hold up other admins trying to rename users, NIPSA users, reindex shorter time frames, etc: all these smaller jobs will get higher priorities

* Removed the `ES_SYNC_JOB_LIMIT` envvar and setting. This is a Celery task parameter now, which is much nicer.

* Refactored the core queue algorithm to use a `collections.Counter` and a single log message, instead of multiple log messages. Also moved log message strings into constants.

  https://github.com/hypothesis/h/blob/6df62d18704f278fceb9b7f6853aeda2adabc004/h/services/search_index/_queue.py#L70-L108

  The tests are now robust against exact log message string wording, log message order, and the existence or not of other log messages:
 https://github.com/hypothesis/h/blob/6df62d18704f278fceb9b7f6853aeda2adabc004/tests/h/services/search_index/_queue_test.py#L95

* Fixed the logging. Log messages were coming out like this:

  ```
  worker (stderr)      | [2020-10-12 17:36:45,892: INFO/ForkPoolWorker-4] {'Annotations synced because they were not in Elasticsearch': 100}
  ```

  Which isn't filterable in Papertrail. Changed it to use Celery's `get_task_logger()`:

  https://github.com/hypothesis/h/blob/7bcb9a075694e1d255997e958f8779ee6ba41403/h/services/search_index/_queue.py#L4-L10

  So they now come out like this:

  ```
  worker (stderr)      | [2020-10-12 17:41:24,378: INFO/ForkPoolWorker-7] h.tasks.indexer.sync_annotations[3f149049-b681-413b-94dd-cb7fac48db5c]: {'Annotations synced because they were not in Elasticsearch': 98, 'Jobs deleted because annotations were up to date in Elasticsearch': 2}
   ```

  This means you can filter on `h.tasks.indexer.sync_annotations` to get all of the `sync_annotations` task's log messages in Papertrail.
